idaes/core/util/model_diagnostics.py

```diff
@@ -500,64 +493,112 @@ def _find_candidate_eqs(candidates_milp, solver, eq_con_list=None, tee=False):
                     candidate_eqns.append(i)
-
+                    
             return candidate_eqns, ds_
         else:
             return None, None
-
-    def svd_analysis(self, n_smallest_sv=10):
+        
+    
+    def svd_analysis(self, n_smallest_sv=10, dense=False):
         """
         Perform SVD analysis of the constraint Jacobian
-
+        
         Args:
             n_smallest_sv: number of smallest singular values to compute
-
+            
         Returns:
             Nothing
-
+            
         Actions:
             Stores SVD results in object
-
+        
         """
-
+        
         if self.n_eq > 1:
-
+        
             # Determine the number of singular values to compute
             # The "-1" is needed to avoid an error with svds
             n_sv = min(n_smallest_sv, min(self.n_eq, self.n_var) - 1)
             print("Computing the", n_sv, "smallest singular value(s)")
-
+        
             # Perform SVD
             # Recall J is a n_eq x n_var matrix
             # Thus U is a n_eq x n_eq matrix
             # And V is a n_var x n_var
             # (U or V may be smaller in economy mode)
             # Thus we really only care about U
-            u, s, v = svds(self.jac_eq, k=n_sv, which="SM")
-
+            
+            if dense:
+                u, s, vT = svd(self.jac_eq.todense(),full_matrices=False)
+                u = np.flip(u[:,-n_sv:], axis=1)
+                s = np.flip(s[-n_sv:], axis=0)
+                vT = np.flip(vT[-n_sv:,:], axis=0)
+            else:
+                # svds does not guarantee the order in which it generates
+                # singular values, but typically generates them least-to-greatest.
+                # Maybe the warning is for singular values of nearly equal
+                # magnitude or a similar edge case?
+                u, s, vT = svds(self.jac_eq, k = n_sv, which='SM')#, solver='lobpcg')
+            
             # Save results
             self.u = u
             self.s = s
-            self.v = v
-
+            self.v = vT.transpose()
+            
         else:
             print(
                 "Warning: model must contain at least 2 equality constraints to perform svd_analysis"
             )
 
+    def underdetermined_variables_and_constraints(self, n_calc=1, tol=0.1, dense=False):
+        """
+        Determines constraints and variables associated with the smallest
+        singular values by having large components in the left and right
+        singular vectors, respectively, associated with those singular values.
+        
+        Args:
+            n_calc: The singular value, as ordered from least to greatest
+            starting from 1, to calculate associated constraints and variables
+            tol: Size below which to ignore constraints and variables in
+            the singular vector
+
+        Returns
+        -------
+        None.
+
+        """
+        if self.s is None:
```

Comment:
Why do we also compute the 10 smallest? What happens if you apply to a small problem with less than 10 variables or constraints?

Reply:
Choosing to compute 10 is somewhat arbitrary, but was the default elsewhere. Applying it to a problem with fewer than 10 variables worked just fine (and is in fact tested), but risked a slightly confusing error message telling the user to rerun `svd_analysis` if they asked for the fifth or sixth singular values in a 5x5 problem. I overhauled the error messages to avoid that slightly confusing situation. As a result, if the user explicitly asks for more singular values than svds can give it, a `ValueError` is raised instead of just computing as many singular values as possible.